<launch>

  <arg name="sensorOffsetX" default="0.0"/>   <!-- 传感器坐标系与车体中心的偏移量 -->
  <arg name="sensorOffsetY" default="0"/>   <!-- 传感器坐标系与车体中心的偏移量 -->
  <arg name="cameraOffsetZ" default="0"/> 
  <arg name="twoWayDrive" default="true"/>  <!-- 双向驱动：就是车体能不能反向走（倒车） -->
  <arg name="maxSpeed" default="1.0"/>      <!-- X、Y轴的最大速度 -->
  <arg name="autonomyMode" default="true"/> <!-- 自主模式，就是没有手柄控制下的模式 -->
  <arg name="autonomySpeed" default="2.0"/>   <!-- 自动探索最大速度 -->
  <arg name="joyToSpeedDelay" default="2.0"/>   
  <arg name="goalX" default="0"/> <!-- 局部路径目标点 -->
  <arg name="goalY" default="0"/> <!-- 局部路径目标点 -->

  <node pkg="local_planner" type="PYH_localPlanner" name="localPlanner" output="screen" required="true">
    <param name="pathFolder" type="string" value="$(find local_planner)/paths" />
    <param name="vehicleLength" type="double" value="1" />      <!-- 车长 单位m-->
    <param name="vehicleWidth" type="double" value="1" />       <!-- 车宽 单位m-->
    <param name="sensorOffsetX" value="$(arg sensorOffsetX)" /> 
    <param name="sensorOffsetY" value="$(arg sensorOffsetY)" />
    <param name="twoWayDrive" value="$(arg twoWayDrive)" />
    <param name="laserVoxelSize" type="double" value="0.05" />  <!-- 下采样体素栅格叶大小 -->
    <param name="terrainVoxelSize" type="double" value="0.2" /> <!-- 下采样体素栅格叶大小 -->
    <param name="useTerrainAnalysis" type="bool" value="true" />  <!--  在室内结构化平坦场地时不使用地形分析的点云使用slam配准后的scan_registered也可以，不过最好使用地形分析的-->
    <param name="checkObstacle" type="bool" value="true" />
    <param name="checkRotObstacle" type="bool" value="false" />   <!-- 没有啥太大作用，待研究 -->
    <param name="adjacentRange" type="double" value="3.25" />       <!-- ①点云距离阈值，受速度影响，以小车为坐标系原点的点云距离在这个阈值内才有效②效果和1一样，都是小车坐标系下的点云 -->
    <param name="obstacleHeightThre" type="double" value="0.15" />  <!-- ①0.15障碍物高度阈值，小于这个值就会被认定成障碍物，否则坡度合适的情况下会认定为斜坡，不会影响far边界的建立,但会影响他局部路径（黄色扇形） -->
    <param name="groundHeightThre" type="double" value="0.01" />     <!-- 0.1地面高度阈值 -->
    <param name="costHeightThre" type="double" value="0.1" />       <!-- 计算路径惩罚得分的权重 -->
    <param name="costScore" type="double" value="0.02" />           <!-- 最小惩罚得分 -->
    <param name="useCost" type="bool" value="false" />
    <param name="pointPerPathThre" type="int" value="2" />          <!-- 判断被筛除路径的一个阈值，不动他 -->
    <param name="minRelZ" type="double" value="-0.5" />             <!-- minRelZ以及maxRelZ限制了有效点的Z值范围，只有不使用TerrainAnalysis的点云才有作用，应该是用于scan_registered的 -->
    <param name="maxRelZ" type="double" value="0.25" />             <!--  -->
    <param name="maxSpeed" value="$(arg maxSpeed)" />
    <param name="dirWeight" type="double" value="0.02" />           <!-- 计算得分时转向角度的权重0.02 -->
    <param name="dirThre" type="double" value="180.0" />             <!-- 最大转向角度90,正负90度，加反向移动，就是360 -->
    <param name="dirToVehicle" type="bool" value="false" />         <!-- dirToVehicle这个参数是为一些类似汽车的机器人平台设置的，这些平台不能原地转向而只能一边向前或者向后运动一边转向。当这个参数设置成false的时候，就是应用在大多数可以差速驱动的机器人平台上，只考虑目标方向附近的轨迹，机器人可以任意无碰撞的转向需要的方向。而当这个参数设置成true的时候，就是应用在上述类似汽车的平台上，这时候只考虑机器人朝向dirThr之内的轨迹，因为机器人只能向前或者向后运动的时候转向，dirThr之外的轨迹机器人是没有能力实现的。一般dirToVehicle是true的时候，dirThr会设置的比较小，比如只考虑机器人朝前或者朝后10°的范围内的轨迹。如果用的是差速驱动的机器人，那就可以保持现在的v默认设置。 -->
    <param name="pathScale" type="double" value="1.25" />           <!-- 路径尺度1.25 比例尺，不要随便改-->
    <param name="minPathScale" type="double" value="0.75" />        <!-- 最小路径尺度0.75 -->
    <param name="pathScaleStep" type="double" value="0.25" />       <!-- 路径尺度的调整步长 -->
    <param name="pathScaleBySpeed" type="bool" value="true" />      <!-- 是否根据速度调整路径尺度 -->
    <param name="minPathRange" type="double" value="1.0" />         <!-- 最小路径距离 -->
    <param name="pathRangeStep" type="double" value="0.5" />        <!-- 路径范围的调整步长 -->
    <param name="pathRangeBySpeed" type="bool" value="true" />      <!-- 是否根据速度调整路径的范围 -->
    <param name="pathCropByGoal" type="bool" value="true" />        <!-- 是否根据目标点+ goalClearRange 筛选点云数据 -->
    <param name="autonomyMode" value="$(arg autonomyMode)" />
    <param name="autonomySpeed" value="$(arg autonomySpeed)" />
    <param name="joyToSpeedDelay" value="$(arg joyToSpeedDelay)" />
    <param name="joyToCheckObstacleDelay" type="double" value="5.0" />
    <param name="goalClearRange" type="double" value="0.5" />       <!-- 当 pathCropByGoal = true 时,点云距离超过目标点+该值则不被处理，不要随便改 -->
    <param name="goalX" type="double" value="$(arg goalX)" />
    <param name="goalY" type="double" value="$(arg goalY)" />
  </node>
  <node pkg="local_planner" type="pathFollower" name="pathFollower" output="screen" required="true">
    <param name="sensorOffsetX" value="$(arg sensorOffsetX)" />
    <param name="sensorOffsetY" value="$(arg sensorOffsetY)" />
    <param name="pubSkipNum" type="int" value="1" />
    <param name="twoWayDrive" value="$(arg twoWayDrive)" />
    <param name="lookAheadDis" type="double" value="0.5" />
    <param name="yawRateGain" type="double" value="7.5" />
    <param name="stopYawRateGain" type="double" value="7.5" />
    <param name="maxYawRate" type="double" value="90.0" />      <!-- 最大角速度（单位：°） -->
    <param name="maxSpeed" value="$(arg maxSpeed)" />
    <param name="maxAccel" type="double" value="1.5" />         <!-- 最大加速度 -->
    <param name="switchTimeThre" type="double" value="1.0" />
    <param name="dirDiffThre" type="double" value="0.1" />
    <param name="stopDisThre" type="double" value="0.2" />
    <param name="slowDwnDisThre" type="double" value="0.85" />
    <param name="useInclRateToSlow" type="bool" value="false" />
    <param name="inclRateThre" type="double" value="120.0" />
    <param name="slowRate1" type="double" value="0.25" />
    <param name="slowRate2" type="double" value="0.5" />
    <param name="slowTime1" type="double" value="2.0" />
    <param name="slowTime2" type="double" value="2.0" />
    <param name="useInclToStop" type="bool" value="false" />
    <param name="inclThre" type="double" value="45.0" />
    <param name="stopTime" type="double" value="5.0" />
    <param name="noRotAtStop" type="bool" value="false" />
    <param name="noRotAtGoal" type="bool" value="true" />
    <param name="autonomyMode" value="$(arg autonomyMode)" />
    <param name="autonomySpeed" value="$(arg autonomySpeed)" />
    <param name="joyToSpeedDelay" value="$(arg joyToSpeedDelay)" />
  </node>

  
 

  

  <node pkg="tf" type="static_transform_publisher" name="vehicleTransPublisher" args="-$(arg sensorOffsetX) -$(arg sensorOffsetY) 0 0 0 0 /sensor /vehicle 1000"/>

  <node pkg="tf" type="static_transform_publisher" name="sensorTransPublisher" args="0 0 $(arg cameraOffsetZ) -1.5707963 0 -1.5707963 /sensor /camera 1000"/>

</launch>
